src/cryptography/hazmat/primitives/ciphers/base.py

```diff
@@ -64,23 +83,59 @@ def tag(self) -> bytes:
         """
 
 
-class Cipher(object):
+Mode = typing.TypeVar(
+    "Mode", bound=typing.Optional[modes.Mode], covariant=True
+)
+
+
+class Cipher(typing.Generic[Mode]):
     def __init__(
         self,
         algorithm: CipherAlgorithm,
-        mode: typing.Optional[modes.Mode],
+        mode: Mode,
         backend: typing.Any = None,
     ):
 
         if not isinstance(algorithm, CipherAlgorithm):
             raise TypeError("Expected interface of CipherAlgorithm.")
 
         if mode is not None:
+            # mypy needs this assert to narrow the type from our generic
+            # type. Maybe it won't some time in the future.
+            assert isinstance(mode, modes.Mode)
             mode.validate_for_algorithm(algorithm)
 
         self.algorithm = algorithm
         self.mode = mode
 
+    @typing.overload
+    def encryptor(self: "Cipher[modes.ModeWithNonce]") -> CipherContext:
```

Comment:
Can this be a union of all the values that would work? either `Cipher[Union[...]]` or `Union[Cipher[X], Cipher[Y], ...]` to reduce verbosity?